app/src/main/java/com/example/next_show/data/ShowFilter.java

```diff
@@ -0,0 +1,169 @@
+    }
+
+    public void removeFromGenre(String g) {
+        filters.get(GENRE).remove(g);
+    }
+
+    public void removeFromNetwork(String n) {
+        filters.get(NETWORK).remove(n);
+    }
+
+    public void removeFromYear(String y) {
+        filters.get(YEAR).remove(y);
+    }
+
+    public void updateShowsList(List<Show> newShows) {
+        showsList.addAll(newShows);
+    }
+
+    public void setShowsList(List<Show> newShows) {
+        allShows = newShows;
+    }
+
+    public List<Show> getAllShows() {
+        return allShows;
+    }
+
+    public List<Show> filterShows(List<Show> shows) {
+        List<Show> filteredShows = shows;
+
+        // filter by network (if appplicable)
+        List<String> networkQueries = filters.get(NETWORK);
+        filteredShows = filterNetwork(filteredShows, networkQueries);
+
+        // filter by genre (if appplicable)
+        List<String> genreQueries = filters.get(GENRE);
+        filteredShows = filterGenre(filteredShows, genreQueries);
+
+        // filter by year (if appplicable)
+        List<String> yearQueries = filters.get(YEAR);
+        filteredShows = filterYear(filteredShows, yearQueries);
+
+        return filteredShows;
+    }
+
+    private List<Show> filterYear(List<Show> current, List<String> queries) {
+        if (queries.isEmpty()) {
+            return current;
+        }
+
+        List<Show> filterByYear = new ArrayList<>();
+        for (Show s : current) {
+            int yearAired = s.getYearAired();
+            if (yearAired >= CURRENT_YEAR && queries.contains(THIS_YEAR)) {
+                filterByYear.add(s);
+            } else if (yearAired <= PAST_YEAR && yearAired > PRIOR_YEAR && queries.contains(PAST_FIVE)) {
+                filterByYear.add(s);
+            } else if (yearAired <= PRIOR_YEAR && queries.contains(PRIOR)) {
+                filterByYear.add(s);
+            }
+        }
+        return filterByYear;
+    }
+
+    private List<Show> filterNetwork(List<Show> current, List<String> queries) {
+        if (queries.isEmpty()) {
+            return current;
+        }
+
+        List<Show> filterByNetwork = new ArrayList<>();
+        for (Show s: current) {
+            String showNetwork = s.getNetwork();
+
+            // if any network match the queries
+            if (queries.contains(STREAMING) && STREAMING_SERVICES.contains(showNetwork)) {
+                filterByNetwork.add(s);
+            } else if (queries.contains(NETWORK) && !STREAMING_SERVICES.contains(showNetwork)) {
+                filterByNetwork.add(s);
+            }
+        }
+        return filterByNetwork;
+
+    }
+
+    public List<Show> filterGenre(List<Show> current, List<String> queries) {
+        if (queries.isEmpty()) {
+            return current;
+        }
+
+        // reuse code from recClient to determine intersection of genres
+        List<Show> filterByGenre = RecommendationClient.getGenreMatch(current, queries);
+        return filterByGenre;
+    }
+}
```

Comment:
This definitely works!  We're going to ender up iterating through the list every single time we want to apply each filter type.  Can you think of a way where we just iterate through the list once and check that it matches the expected filters?

Reply:
(Also feel free to chat about the solution and then once we're happy with the it, you can then implement an optimized version)